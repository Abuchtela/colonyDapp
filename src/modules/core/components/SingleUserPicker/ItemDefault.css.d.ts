export const main: string;
export const dataContainer: string;
export const displayName: string;
export const showAddress: string;
export const fullName: string;
export const address: string;
export const thatsYou: string;
export const mention: string;
