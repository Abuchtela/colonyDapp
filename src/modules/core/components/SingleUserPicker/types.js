/* @flow */

export type UserData = {
  id: string,
  fullName?: string,
  username?: string,
};
