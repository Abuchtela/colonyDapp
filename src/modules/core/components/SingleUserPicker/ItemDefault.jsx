/* @flow */

import React from 'react';
import { defineMessages, FormattedMessage } from 'react-intl';
import cx from 'classnames';

import type { UserData } from './types';

import UserAvatar from '../UserAvatar';

import styles from './ItemDefault.css';

const MSG = defineMessages({
  ownName: {
    id: 'SingleUserPicker.ItemWithYouText.youText',
    defaultMessage: '(you)',
  },
});

type Props = {
  currentUserId?: string,
  itemData: UserData,
  selected?: boolean,
  showAddress?: boolean,
};

const ItemDefault = ({
  currentUserId,
  itemData: { id, fullName, username },
  showAddress,
}: Props) => (
  <span className={cx(styles.main, { [styles.showAddress]: showAddress })}>
    <UserAvatar size="s" userId={id} username={username || id} />
    <span className={styles.dataContainer}>
      {fullName && (
        <span className={styles.fullName}>
          {fullName}
          {currentUserId === id && (
            <span className={styles.thatsYou}>
              <FormattedMessage {...MSG.ownName} />
            </span>
          )}
        </span>
      )}
      {username && <span className={styles.username}>{`@${username}`}</span>}
      {showAddress && <span className={styles.address}>{id}</span>}
    </span>
  </span>
);

ItemDefault.displayName = 'SingleUserPicker.ItemDefault';

export default ItemDefault;
