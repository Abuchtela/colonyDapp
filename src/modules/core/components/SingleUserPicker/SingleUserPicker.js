// @flow
import { asField } from '../Fields';

import SingleUserPicker from './SingleUserPicker.jsx';

export default asField()(SingleUserPicker);
