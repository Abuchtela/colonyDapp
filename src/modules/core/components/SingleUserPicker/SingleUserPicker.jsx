/* @flow */

import type { MessageDescriptor } from 'react-intl';

import React, { Component } from 'react';
import { defineMessages } from 'react-intl';
import { compose } from 'recompose';

import styles from './SingleUserPicker.css';

import { getMainClasses } from '~utils/css';

import type { OmniPickerProps, OmniPickerItemComponent } from '../OmniPicker';

import { asField, InputLabel } from '../Fields';
import Icon from '../Icon';
import { withOmniPicker } from '../OmniPicker';
// import UserAvatar from '../../../users/components/UserAvatar';

type UserData = {
  id: string,
  fullName: string,
};

const MSG = defineMessages({
  selectMember: {
    id: 'SingleUserPicker.selectMember',
    defaultMessage: 'Select member',
  },
  emptyMessage: {
    id: 'SingleUserPicker.emptyMessage',
    defaultMessage: 'No Colony members match that search.',
  },
});

// TODO: Replace with proper UserAvatar component
const UserAvatar = () => (
  <Icon
    className={styles.focusIcon}
    name="circle-person"
    title={MSG.selectMember}
  />
);

type Appearance = {
  direction?: 'horizontal',
};

type Props = {
  /** Appearance object */
  appearance?: Appearance,
  /** Connect to form state (will inject `$value`, `$id`, `$error`, `$touched`), is `true` by default */
  connect?: boolean,
  /** Just render the `<textarea>` element without label */
  elementOnly?: boolean,
  /** Textarea field name (form variable) */
  name: string,
  /** Help text (will appear next to label text) */
  help?: string | MessageDescriptor,
  /** Values for help text (react-intl interpolation) */
  helpValues?: { [string]: string },
  /** Item component for omnipicker listbox */
  itemComponent: OmniPickerItemComponent,
  /** Label text */
  label: string | MessageDescriptor,
  /** Values for label text (react-intl interpolation) */
  labelValues?: { [string]: string },
  /** Placeholder for input */
  placeholder?: string,
  /** @ignore Will be injected by `asField` */
  $error?: string,
  /** @ignore Will be injected by `asField` */
  $value?: string,
  /** @ignore Will be injected by `asField` */
  $touched?: boolean,
  /** @ignore Will be injected by `asField` */
  setValue: (val: any) => void,
} & OmniPickerProps;

type State = {
  selectedUser: ?UserData,
};

class SingleUserPicker extends Component<Props, State> {
  static displayName = 'SingleUserPicker';

  state = {
    selectedUser: null,
  };

  handleActiveUserClick = () => {
    const { inputNode } = this.props;
    this.setState(
      {
        selectedUser: null,
      },
      () => {
        if (inputNode) inputNode.focus();
      },
    );
  };

  handlePick = (user: UserData) => {
    this.setState({
      selectedUser: user,
    });
  };

  render() {
    const {
      appearance,
      itemComponent,
      // Form field
      $error,
      $touched,
      elementOnly,
      help,
      label,
      placeholder,
      // OmniPicker
      inputProps,
      OmniPicker,
      OmniPickerWrapper,
      omniPickerIsOpen,
      registerInputNode,
    } = this.props;

    const { selectedUser } = this.state;
    const labelAppearance = appearance
      ? { direction: appearance.direction }
      : undefined;
    return (
      <OmniPickerWrapper className={getMainClasses(appearance, styles)}>
        <div className={styles.inputContainer}>
          {!elementOnly &&
            label && (
              <InputLabel
                inputId={inputProps.id}
                label={label}
                help={help}
                error={$error}
                appearance={labelAppearance}
              />
            )}
          {selectedUser ? (
            <div className={styles.avatarContainer}>
              <UserAvatar
                className={styles.recipientAvatar}
                userId={selectedUser.id}
                size="xs"
              />
            </div>
          ) : (
            <Icon
              className={omniPickerIsOpen ? styles.focusIcon : styles.icon}
              name="circle-person"
              title={MSG.selectMember}
            />
          )}
          {}
          <div className={styles.container}>
            {/* eslint-disable jsx-a11y/click-events-have-key-events */
            selectedUser && (
              <div
                role="button"
                className={styles.recipientName}
                onClick={this.handleActiveUserClick}
                onFocus={this.handleActiveUserClick}
                tabIndex="0"
              >
                {selectedUser.fullName}
              </div>
            )}
            {/* eslint-enable jsx-a11y/click-events-have-key-events */}
            <input
              className={
                $touched && $error ? styles.inputInvalid : styles.input
              }
              {...inputProps}
              placeholder={placeholder}
              hidden={selectedUser}
              ref={registerInputNode}
            />
            {$error &&
              appearance &&
              appearance.direction === 'horizontal' && (
                <span className={styles.errorHorizontal}>{$error}</span>
              )}
            <div className={styles.omniPickerContainer}>
              <OmniPicker
                itemComponent={itemComponent}
                onPick={this.handlePick}
              />
            </div>
          </div>
        </div>
      </OmniPickerWrapper>
    );
  }
}

const enhance = compose(
  asField(),
  withOmniPicker(),
);

export default enhance(SingleUserPicker);
