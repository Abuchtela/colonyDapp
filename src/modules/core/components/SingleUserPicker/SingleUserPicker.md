```

const Item = ({ itemData: { username, name }, selected }) => (
  <div style={{ backgroundColor: selected ? '#e7e8e8' : 'transparent', padding: '20px'}}>
    Username: {username}<br />
    Name: {name}
  </div>
);

const data = [
  { id: 1, username: 'Chris', fullName: 'Christian Maniewski'},
  { id: 2, username: 'Elena', fullName: 'Elena Dimitrova'},
  { id: 3, username: 'Thiago', fullName: 'Thiago Delgado'},
  { id: 4, username: 'Alex', fullName: 'Alex Rea'},
  { id: 5, username: 'Collin', fullName: 'Collin Vine'},
];

const filter = (data, filterValue) => data.filter(user => user.username.toLowerCase().startsWith(filterValue.toLowerCase()));

<SingleUserPicker label="Pick user" connect={false} itemComponent={Item} data={data} filter={filter} />
```
