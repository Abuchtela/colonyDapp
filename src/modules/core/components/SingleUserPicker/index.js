/* @flow */

export { default } from './SingleUserPicker.jsx';

export { default as ItemDefault } from './ItemDefault.jsx';

export type * from './types';
