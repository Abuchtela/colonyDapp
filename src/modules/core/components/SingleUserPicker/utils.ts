import { AnyUser } from '~data/index';

/*
Extracts the required values to be used in the SingleUserPicker
on selection
*/
export const filterUserSelection = (data: AnyUser[], filterValue: string) => {
  if (!filterValue) {
    return data;
  }

  const filtered = data.filter((user) => {
    const { username, walletAddress } = user?.profile;
    const valueToLowerCase = filterValue.toLowerCase();
    return (
      username &&
      walletAddress &&
      (`@${username}`.toLowerCase().includes(valueToLowerCase) ||
        walletAddress.toLowerCase().includes(valueToLowerCase))
    );
  });

  const customUserValue: AnyUser = {
    id: 'filterValue',
    profile: {
      walletAddress: filterValue,
      displayName: filterValue,
    },
  };

  return [customUserValue].concat(filtered);
};
