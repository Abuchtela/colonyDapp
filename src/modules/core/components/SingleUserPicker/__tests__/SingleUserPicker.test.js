/* eslint-env jest */

import React from 'react';
import toJson from 'enzyme-to-json';

import { shallowWithIntl, mountWithIntlContext, fieldPropsFactory } from 'testutils';

import SingleUserPicker from '../SingleUserPicker.jsx';
import { UserInfoPopover } from '../../UserInfo';
import { ESC } from '../../../../keyTypes';

const createFieldProps = fieldPropsFactory({
  selector: (_, { data }) => data,
  userInfoTypes: [],
  itemComponent: () => null,
  users: [
    {
      _id: '1',
      profile: {
        username: 'thiago',
        fullName: 'Thiago D.',
        address: 'thiagoxaddress',
      },
    },
    {
      _id: '2',
      profile: {
        username: 'chris',
        fullName: 'Christian M.',
        address: 'chrisxaddress',
      },
    },
  ],
}, {
  utils: {
    getIntlFormatted: jest.fn(),
  },
  elementOnly: true,
  label: 'recipient',
},
);

jest.mock('../../../../users/components/UserAvatar', () => () => null);

describe('SingleUserPicker component', () => {
  test('Renders initial component', () => {
    const wrapper = shallowWithIntl((
      <SingleUserPicker
        {...createFieldProps()}
      />
    ));
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('User selector open on focus', () => {
    const onChangeSpy = jest.fn();
    const wrapper = mountWithIntlContext((
      <SingleUserPicker
        {...createFieldProps({ onChange: onChangeSpy })}
      />
    ));
    wrapper.instance().handleFocus();
    expect(wrapper).toHaveState('omniPickerOpen', true);
    expect(wrapper.html()).toContain('data-enzyme-hook="omni-picker"');
  });

  test('ESC closes user selector, always', () => {
    const onChangeSpy = jest.fn();
    const wrapper = mountWithIntlContext((
      <SingleUserPicker
        {...createFieldProps({ onChange: onChangeSpy })}
      />
    ));
    const input = wrapper.find('input');
    wrapper.instance().handleFocus();
    expect(wrapper).toHaveState('omniPickerOpen', true);
    input.simulate('keyup', { key: ESC });
    expect(wrapper).toHaveState('omniPickerOpen', false);
  });

  test('If user selected has Avatar (it tests the UserInfoPopover wrapper as the Avatar is mocked)', () => {
    const currentSelectedUser = {
      _id: '1',
      profile: {
        username: 'thiago',
        fullName: 'Thiago D.',
        address: 'thiagoxaddress',
      },
    };
    const onChangeSpy = jest.fn();

    const wrapper = mountWithIntlContext((
      <SingleUserPicker
        {...createFieldProps({ onChange: onChangeSpy })}
      />
    ));
    wrapper.setState({ selectedUser: currentSelectedUser });
    expect(wrapper.find(UserInfoPopover)).toBePresent();
  });


  test('If user selected input is hidden', () => {
    const currentSelectedUser = {
      _id: '1',
      profile: {
        username: 'thiago',
        fullName: 'Thiago D.',
        address: 'thiagoxaddress',
      },
    };
    const onChangeSpy = jest.fn();

    const wrapper = mountWithIntlContext((
      <SingleUserPicker
        {...createFieldProps({ onChange: onChangeSpy })}
      />
    ));
    wrapper.setState({ selectedUser: currentSelectedUser });
    expect(wrapper.find('input[hidden]')).toBePresent();
  });
});
