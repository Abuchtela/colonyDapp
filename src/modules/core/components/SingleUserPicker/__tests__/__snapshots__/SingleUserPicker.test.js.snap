// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleUserPicker component Renders initial component 1`] = `
<div
  className=""
>
  <div>
    <InjectIntl(withProps(core.Icon))
      name="circle-person"
    />
    <div>
      <input
        autoComplete="off"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        value=""
      />
    </div>
  </div>
</div>
`;
